src/server/pingserver-rs/src/admin.rs

```diff
@@ -104,33 +105,61 @@ impl Admin {
-                            let mut session = Session::new(
-                                addr,
-                                stream,
-                                State::Handshaking,
-                                Some(rustls::ServerSession::new(&tls_config)),
-                                self.metrics.clone(),
-                            );
-                            let s = self.sessions.vacant_entry();
-                            let token = s.key();
-                            session.set_token(Token(token));
-                            let _ = session.register(&self.poll);
-                            s.insert(session);
+                        // handle TLS if it is configured
+                        if let Some(ssl_context) = &self.ssl_context {
+                            match Ssl::new(&ssl_context).map(|v| v.accept(stream)) {
+                                // handle case where we have a fully-negotiated
+                                // TLS stream on accept()
+                                Ok(Ok(tls_stream)) => {
+                                    let mut session =
+                                        Session::tls(addr, tls_stream, self.metrics.clone());
+                                    trace!("accepted new session: {}", addr);
+                                    let s = self.sessions.vacant_entry();
+                                    let token = s.key();
+                                    session.set_token(Token(token));
+                                    if session.register(&self.poll).is_ok() {
+                                        s.insert(session);
+                                    } else {
+                                        let _ =
+                                            self.metrics().increment_counter(&Stat::TcpAcceptEx, 1);
+                                    }
+                                }
+                                // handle case where further negotiation is
+                                // needed
+                                Ok(Err(HandshakeError::WouldBlock(tls_stream))) => {
+                                    let mut session = Session::handshaking(
+                                        addr,
+                                        tls_stream,
+                                        self.metrics.clone(),
+                                    );
+                                    let s = self.sessions.vacant_entry();
+                                    let token = s.key();
+                                    session.set_token(Token(token));
+                                    if session.register(&self.poll).is_ok() {
+                                        s.insert(session);
+                                    } else {
+                                        let _ =
+                                            self.metrics().increment_counter(&Stat::TcpAcceptEx, 1);
+                                    }
```

Comment:
line 135-143 is a duplicate of 117-125? Can we shuffle the order of these cases and use fall through to avoid duplication?

Reply:
Fixing...

Reply:
nevermind - this is already fixed =)